pkg/proxy/iptables/proxier.go

```diff
@@ -0,0 +1,567 @@
+		}
+	}
+}
+
+// used in OnEndpointsUpdate
+type hostPortProtocol struct {
+	host     string
+	port     int
+	protocol api.Protocol
+}
+
+func isValidEndpoint(hpp *hostPortProtocol) bool {
+	return hpp.host != "" && hpp.port > 0
+}
+
+// Tests whether two slices are equivalent.  This sorts both slices in-place.
+func slicesEquiv(lhs, rhs []string) bool {
+	if len(lhs) != len(rhs) {
+		return false
+	}
+	if reflect.DeepEqual(slice.SortStrings(lhs), slice.SortStrings(rhs)) {
+		return true
+	}
+	return false
+}
+
+func flattenValidEndpoints(endpoints []hostPortProtocol) []string {
+	// Convert Endpoint objects into strings for easier use later.  Ignore
+	// the protocol field - we'll get that from the Service objects.
+	var result []string
+	for i := range endpoints {
+		hpp := &endpoints[i]
+		if isValidEndpoint(hpp) {
+			result = append(result, net.JoinHostPort(hpp.host, strconv.Itoa(hpp.port)))
+		}
+	}
+	return result
+}
+
+// servicePortToServiceChain takes the ServicePortName for a
+// service and returns the associated iptables chain
+// this is computed by hashing (sha256) then encoding to base64 and
+// truncating with the prefix "KUBE-SVC-"
+// We do this because Iptables Chain Names must be <= 28 chars long
+func servicePortToServiceChain(s proxy.ServicePortName) utiliptables.Chain {
+	hash := sha256.Sum256([]byte(s.String()))
+	encoded := base64.URLEncoding.EncodeToString(hash[:])
+	return utiliptables.Chain("KUBE-SVC-" + encoded[:19])
+}
+
+// this is the same as servicePortToServiceChain but with the endpoint included essentially
+func servicePortAndEndpointToServiceChain(s proxy.ServicePortName, endpoint string) utiliptables.Chain {
+	hash := sha256.Sum256([]byte(s.String() + "_" + endpoint))
+	encoded := base64.URLEncoding.EncodeToString(hash[:])
+	return utiliptables.Chain("KUBE-SVC-" + encoded[:19])
+}
+
+const (
+	fmtChain  = ":%s - [0:0]\n"
+	fmtFlush  = "-F %s\n"
+	fmtDelete = "-X %s\n"
+)
+
+// This is where all of the iptables-save/restore calls happen.
+// The only other iptables rules are those that are setup in iptablesInit()
+func (proxier *Proxier) syncProxyRules() error {
+	glog.V(2).Infof("Syncing iptables rules.")
+
+	// lock proxy state so that we can sync rules from the current state
+	proxier.mu.Lock()
+	defer proxier.mu.Unlock()
+
+	// Get iptables-save output so we can check for existing chains and rules.
+	// This will be a map of chain name to chain with rules as stored in iptables-save/iptables-restore
+	existingChains := make(map[utiliptables.Chain]string)
+	// run iptables-save
+	iptablesSaveRaw, err := proxier.iptables.Save(utiliptables.TableNAT)
+	if err != nil { // if we failed to get any rules
+		glog.Errorf("Failed to execute iptable-save, syncing all rules. (%s)", err.Error())
+	} else { // otherwise parse the output
+		existingChains = utiliptables.ParseTableSaveChains(utiliptables.TableNAT, iptablesSaveRaw)
+	}
+
+	// for first line and chains
+	var rules bytes.Buffer
```

Comment:
why not name these "chains" and "rules" or "chainLines" and "ruleLines" or something?